main.go

```diff
@@ -0,0 +1,166 @@
+		level.Error(logger).Log("msg", "add api", "error", err)
+		return tenantCfg, err
+	}
+	tenantCfg.Tenant = tenant
+	tenantCfg.OIDC.Audience = os.Getenv("OIDC_AUDIENCE")
+	tenantCfg.OIDC.ClientID = os.Getenv("OIDC_CLIENT_ID")
+	tenantCfg.OIDC.ClientSecret = os.Getenv("OIDC_CLIENT_SECRET")
+	tenantCfg.OIDC.IssuerURL = os.Getenv("OIDC_ISSUER_URL")
+	if _, err := tenantCfg.Client(ctx, logger); err != nil {
+		level.Error(logger).Log("msg", "creating authenticated client", "error", err)
+		return tenantCfg, err
+	}
+
+	if err := conf.AddTenant(logger, tenantCfg.Tenant, obsctlContext, tenantCfg.Tenant, tenantCfg.OIDC); err != nil {
+		level.Error(logger).Log("msg", "adding tenant", "error", err)
+		return tenantCfg, err
+	}
+
+	return tenantCfg, nil
+}
+
+func ObsctlMetricsSet(ctx context.Context, tenantConfig config.TenantConfig, rules monitoringv1.PrometheusRuleSpec) error {
+	level.Info(logger).Log("msg", "setting metrics for tenant", "tenant", tenantConfig.Tenant)
+	api := os.Getenv("OBSERVATORIUM_URL")
+	c, err := tenantConfig.Client(ctx, logger)
+	if err != nil {
+		level.Error(logger).Log("msg", "getting client", "error", err)
+		return err
+	}
+	fc, err := client.NewClientWithResponses(api, func(f *client.Client) error {
+		f.Client = c
+		return err
+	})
+	if err != nil {
+		level.Error(logger).Log("msg", "getting fetcher client", "error", err)
+		return err
+	}
+	body, err := json.Marshal(rules)
+	if err != nil {
+		level.Error(logger).Log("msg", "converting rules to json", "error", err)
+		return err
+	}
+	resp, err := fc.SetRawRulesWithBodyWithResponse(ctx, parameters.Tenant(tenantConfig.Tenant), "application/yaml", bytes.NewReader(body))
+	if err != nil {
+		level.Error(logger).Log("msg", "getting response", "error", err)
+		return err
+	}
+	if resp.StatusCode()/100 != 2 {
+		if len(resp.Body) != 0 {
+			level.Error(logger).Log("msg", "setting rules", "error", string(resp.Body))
+			return err
+		}
+	}
+	level.Info(logger).Log("msg", string(resp.Body))
+
+	return nil
+}
+
+func Sleep() {
+	sleepDurationSeconds := defaultSleepDurationSeconds
+	if value, ok := os.LookupEnv("SLEEP_DURATION_SECONDS"); ok {
+		sleepDurationSeconds, _ = strconv.Atoi(value)
+	}
+	level.Info(logger).Log("msg", "sleeping", "duration", sleepDurationSeconds)
+	time.Sleep(time.Duration(sleepDurationSeconds) * time.Second)
+}
+
+func main() {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	sigs := make(chan os.Signal, 1)
+	signal.Notify(sigs, os.Interrupt, syscall.SIGTERM)
+	go func() {
+		<-sigs
+		cancel()
+	}()
+
+	logger = log.NewLogfmtLogger(log.NewSyncWriter(os.Stderr))
+	for true {
+		prometheusRules, err := GetPrometheusRules(ctx)
+		if err != nil {
+			level.Error(logger).Log("msg", "error getting prometheus rules")
+			os.Exit(1)
+		}
+		tenantRules := GetTenantRules(prometheusRules)
+		for tenant, rules := range tenantRules {
+			tenantConfig, err := InitObsctlTenantConfig(ctx, tenant)
```

Comment:
This should've been erroring out with ["api with name api already exists"](https://github.com/observatorium/obsctl/blob/main/pkg/config/config.go#L232) as this loop continuously adds already existing API/tenant. But seeing that this is already deployed and working, I tried to dig a bit deeper,

So `InitObsctlTenantConfig` does not read the previous config from disk and instead creates a new one each time. This also effectively means that it requests a token from SSO after each iteration (after 30s or so), regardless of token validity (usually ~15m). 

Maybe this can be improved later on so that if the details of a tenant are already saved, it would not need to initialize `tenantCfg` again and reuse SSO token until it expires (this is already [there](https://github.com/observatorium/obsctl/blob/main/pkg/config/config.go#L113) in obsctl)? 🙂

Reply:
i agree this would be a significant improvement. i also appreciate the approach of doing this in a future PR.
i am a big believer in:
- make it work <-- we are here
- make it work right
- make it beautiful